cluster.go

```diff
@@ -57,45 +59,95 @@ func startServer(
 			if timeElapsed.Milliseconds() > ElectionTimeOut {
 				isElection = true
 				timeSinceLastUpdate = time.Now()
-				go elect(&state, voteChannels, leaderCommunicationChannels)
+				go elect(&state, voteChannels, onWinChannel)
 			}
 			time.Sleep(electionThreadSleepTime)
 		}
 	}()
 
-	// receive messages from leader
+	/* Handles messages from leader. Duties include:
+	* - ignore anything with stale term
+	 * - update timeSinceLastUpdate
+	 * - process new log entries + heartbeats (empty logs)
+	 */
 	go func () {
 		for newLogEntry := range leaderCommunicationChannels[state.ServerId] {
-			if isElection { //received message from leader during election,
-				serverStateLock.Lock()
-				state.Role = FollowerRole // for candidates that lost the election
-				serverStateLock.Unlock()
-				isElection = false
+			 if newLogEntry.Term >= state.CurrentTerm {
+				timeSinceLastUpdate = time.Now()
+
+				if isElection { //received message from leader during election,
+					isElection = false
+					serverStateLock.Lock()
+					if state.Role != LeaderRole {
+						state.Role = FollowerRole // for candidates that lost the election
+					}
+					serverStateLock.Unlock()
+				}
+
+				printMessageFromLeader(state.ServerId, newLogEntry)
+				//process log entry here
 			}
-			timeSinceLastUpdate = time.Now()
-			fmt.Print("New log entry received from leader:", newLogEntry, "\n")
-			//process log entry here
+
 		}
 		done <- true
 	}()
+
+	/* On Win Handler. Responsibilities includes
+	 * - updating role of server to leader
+	 * - running heartbeat thread
+	 * - managing client requests (in progress)
+	 */
+	go func(){
+		select {
+		case <-onWinChannel: // got enough votes
+			serverStateLock.Lock()
+			state.Role = LeaderRole
+			serverStateLock.Unlock()
+			go runHeartbeatThread(&state, leaderCommunicationChannels)
+			go readAndDistributeClientRequests()
+		}
+	}()
+}
+
+func runHeartbeatThread(state * ServerState, leaderCommunicationChannels *[ClusterSize]chan LogEntry) {
+	for state.Role == LeaderRole {
+		for serverIndex, leaderCommunicationChannel := range *leaderCommunicationChannels {
+			leaderCommunicationChannel <- LogEntry{serverIndex, state.CurrentTerm, KeyValue{"", ""}}
+		}
+		fmt.Println() //breaks up prints into chunks for each beat.
+		time.Sleep(HeartBeatDelay * time.Millisecond)
+	}
```

Comment:
- Can we make this function take in a key-value pair so we can use it later for sending data to followers (and just put a blank key value pair in at line 106 for now)?
 - Our ability to do this might depend on how we're supposed to communicate with the client which we can talk to Maria about in office hours tomorrow

Reply:
Aren't heart beat messages supposed to be blank and non-empty one sent whenever the client wants something done? Regardless, we can talk to maria